autogpts/autogpt/autogpt/agents/agent.py

```diff
@@ -277,69 +291,58 @@ async def execute(
-
-        # Update action history
-        self.event_history.register_result(result)
-        await self.event_history.handle_compression(
-            self.llm_provider, self.legacy_config
-        )
+        await self.foreach_components("after_execution", result)
+
+        self.print_trace()
 
         return result
 
+    def print_trace(self):
+        print("\n".join(self.trace))
 
-#############
-# Utilities #
-#############
-
-
-async def execute_command(
-    command_name: str,
-    arguments: dict[str, str],
-    agent: Agent,
-) -> CommandOutput:
-    """Execute the command and return the result
-
-    Args:
-        command_name (str): The name of the command to execute
-        arguments (dict): The arguments for the command
-        agent (Agent): The agent that is executing the command
-
-    Returns:
-        str: The result of the command
-    """
-    # Execute a native command with the same name or alias, if it exists
-    if command := agent.command_registry.get_command(command_name):
-        try:
-            result = command(**arguments, agent=agent)
-            if inspect.isawaitable(result):
-                return await result
-            return result
-        except AgentException:
-            raise
-        except Exception as e:
-            raise CommandExecutionError(str(e))
-
-    # Handle non-native commands (e.g. from plugins)
-    if agent._prompt_scratchpad:
-        for name, command in agent._prompt_scratchpad.commands.items():
-            if (
-                command_name == name
-                or command_name.lower() == command.description.lower()
-            ):
-                try:
-                    return command.method(**arguments)
-                except AgentException:
-                    raise
-                except Exception as e:
-                    raise CommandExecutionError(str(e))
-
-    raise UnknownCommandError(
-        f"Cannot execute command '{command_name}': unknown command."
-    )
+    async def execute_command(
+        self,
+        command_name: str,
+        arguments: dict[str, str],
+    ) -> CommandOutput:
+        """Execute the command and return the result
+
+        Args:
+            command_name (str): The name of the command to execute
+            arguments (dict): The arguments for the command
+
+        Returns:
+            str: The result of the command
+        """
+        # Execute a native command with the same name or alias, if it exists
+        if command := self.get_command(command_name):
+            try:
+                result = command(**arguments)
+                if inspect.isawaitable(result):
+                    return await result
+                return result
+            except AgentException:
+                raise
+            except Exception as e:
+                raise CommandExecutionError(str(e))
+
+        raise UnknownCommandError(
+            f"Cannot execute command '{command_name}': unknown command."
+        )
+
```

Comment:
`self.commands` is a temp solution, once everything becomes one pipeline there'll be no need to cache them in Agent.